tiledb/sm/filter/xor_filter.cc

```diff
@@ -0,0 +1,251 @@
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * This file implements class XORFilter.
+ */
+
+#include "tiledb/common/logger.h"
+#include "tiledb/sm/enums/filter_option.h"
+#include "tiledb/sm/filter/filter_buffer.h"
+#include "tiledb/sm/tile/tile.h"
+
+#include "tiledb/sm/filter/xor_filter.h"
+
+using namespace tiledb::common;
+
+namespace tiledb {
+namespace sm {
+
+void XORFilter::dump(FILE* out) const {
+  if (out == nullptr)
+    out = stdout;
+  fprintf(out, "XORFilter");
+}
+
+Status XORFilter::run_forward(
+    const Tile& tile,
+    Tile* const,  // offsets_tile
+    FilterBuffer* input_metadata,
+    FilterBuffer* input,
+    FilterBuffer* output_metadata,
+    FilterBuffer* output) const {
+  auto tile_type = tile.type();
+
+  switch (datatype_size(tile_type)) {
+    case sizeof(int8_t): {
+      return run_forward<int8_t>(
+          input_metadata, input, output_metadata, output);
+    }
+    case sizeof(int16_t): {
+      return run_forward<int16_t>(
+          input_metadata, input, output_metadata, output);
+    }
+    case sizeof(int32_t): {
+      return run_forward<int32_t>(
+          input_metadata, input, output_metadata, output);
+    }
+    case sizeof(int64_t): {
+      return run_forward<int64_t>(
+          input_metadata, input, output_metadata, output);
+    }
+    default: {
+      return Status_FilterError(
+          "XORFilter::run_forward: datatype size cannot be converted to "
+          "integer type.");
+    }
+  }
+
+  return Status_FilterError("XORFilter::run_forward: invalid datatype.");
+}
+
+template <typename T>
+Status XORFilter::run_forward(
+    FilterBuffer* input_metadata,
+    FilterBuffer* input,
+    FilterBuffer* output_metadata,
+    FilterBuffer* output) const {
+  // Output size does not change with this filter.
+  RETURN_NOT_OK(output->prepend_buffer(input->size()));
+  Buffer* output_buf = output->buffer_ptr(0);
+  assert(output_buf != nullptr);
+
+  // Write the metadata
+  auto parts = input->buffers();
+  auto num_parts = (uint32_t)parts.size();
+  uint32_t metadata_size = sizeof(uint32_t) + num_parts * sizeof(uint32_t);
+  RETURN_NOT_OK(output_metadata->append_view(input_metadata));
+  RETURN_NOT_OK(output_metadata->prepend_buffer(metadata_size));
+  RETURN_NOT_OK(output_metadata->write(&num_parts, sizeof(uint32_t)));
+
+  // Shuffle all parts
+  for (const auto& part : parts) {
+    auto part_size = (uint32_t)part.size();
+    RETURN_NOT_OK(output_metadata->write(&part_size, sizeof(uint32_t)));
+    RETURN_NOT_OK(shuffle_part<T>(&part, output_buf));
+  }
+
+  return Status::Ok();
+}
+
+template <typename T>
+Status XORFilter::shuffle_part(const ConstBuffer* part, Buffer* output) const {
+  uint32_t s = part->size();
+  assert(s % sizeof(T) == 0);
+  uint32_t num_elems_in_part = s / sizeof(T);
+
+  if (num_elems_in_part == 0) {
+    return Status::Ok();
+  }
+
+  // Write starting element.
+  const T* part_array = static_cast<const T*>(part->data());
```

Comment:
We need to cast into an integer of the same width as the input type here, because this needs to support double and float as input.

Reply:
(also add a test in the test section)

Reply:
Yes, this is exactly what we do. I case => template not on the type, but the size of the type.

Reply:
I see, thanks - looks good then. Could you please add:
- `enable_if(is_integral<T>::true)` check on this template so that is _only_ usable for integers
- code comments here and at the switch statement, to clarify why/what is happening

Reply:
I've made those changes.